media/src/main/java/io/wcm/handler/media/spi/MediaSource.java

```diff
@@ -386,27 +390,82 @@ private boolean resolveFirstMatchRenditions(Media media, Asset asset, MediaArgs
+    final List<Rendition> resolvedRenditions = new ArrayList<>();
+
+    List<MediaFormatOption> parentMediaFormatOptions = getParentMediaFormats(mediaArgs);
+    allMandatoryResolved = resolveRenditionsWithMediaFormats(asset, mediaArgs, parentMediaFormatOptions, resolvedRenditions);
+
+    if (allMandatoryResolved && !resolvedRenditions.isEmpty()) {
+      List<MediaFormat> resolvedMediaFormats = resolvedRenditions.stream()
+          .map(Rendition::getMediaFormat)
+          .filter(Objects::nonNull)
+          .collect(Collectors.toList());
+
+      for (MediaFormat mediaFormat : resolvedMediaFormats) {
+        List<MediaFormatOption> childMediaFormatOptions = getChildMediaFormats(mediaArgs, mediaFormat);
+        allMandatoryResolved = allMandatoryResolved && resolveRenditionsWithMediaFormats(asset, mediaArgs, childMediaFormatOptions, resolvedRenditions);
+      }
+    }
+
+    anyResolved = CollectionUtils.isNotEmpty(resolvedRenditions);
+
+    media.setRenditions(resolvedRenditions);
+    if (!resolvedRenditions.isEmpty()) {
+      media.setUrl(resolvedRenditions.get(0).getUrl());
+    }
+    return anyResolved && allMandatoryResolved;
+  }
+
+  private boolean resolveRenditionsWithMediaFormats(@NotNull Asset asset, @NotNull MediaArgs mediaArgs,
+      @NotNull List<MediaFormatOption> mediaFormatOptions, @NotNull List<Rendition> resolvedRenditions) {
+    boolean allMandatoryResolved = true;
+    for (MediaFormatOption mediaFormatOption : mediaFormatOptions) {
       MediaArgs renditionMediaArgs = mediaArgs.clone();
       renditionMediaArgs.mediaFormat(mediaFormatOption.getMediaFormat());
       Rendition rendition = asset.getRendition(renditionMediaArgs);
       if (rendition != null) {
         resolvedRenditions.add(rendition);
-        anyResolved = true;
       }
       else if (mediaFormatOption.isMandatory()) {
         allMandatoryResolved = false;
       }
     }
-    media.setRenditions(resolvedRenditions);
-    if (!resolvedRenditions.isEmpty()) {
-      media.setUrl(resolvedRenditions.get(0).getUrl());
-    }
-    return anyResolved && allMandatoryResolved;
+    return allMandatoryResolved;
+  }
+
+  @NotNull
+  private List<MediaFormatOption> getParentMediaFormats(@NotNull MediaArgs mediaArgs) {
+    return Arrays.stream(mediaArgs.getMediaFormatOptions())
+        .filter(this::isParentMediaFormat)
+        .collect(Collectors.toList());
+  }
+
+  @NotNull
+  private List<MediaFormatOption> getChildMediaFormats(@NotNull MediaArgs mediaArgs, @NotNull final MediaFormat parentMediaFormat) {
+    return Arrays.stream(mediaArgs.getMediaFormatOptions())
+        .filter(this::isChildMediaFormat)
+        .filter(childMediaFormat -> hasParent(childMediaFormat, parentMediaFormat))
+        .collect(Collectors.toList());
+  }
+
+  private boolean isParentMediaFormat(@NotNull MediaFormatOption mediaFormatOption) {
+    return Objects.isNull(getParentMediaFormat(mediaFormatOption.getMediaFormat()));
```

Comment:
getMediaFormat() can be null - leads to null-check warnings (also on the next two methods)

Reply:
fixed